3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,350 @@
+
+  private def extractStrategyFunction(config: JValue): Validation[String, String] =
+    ScalazJson4sUtils
+      .extract[String](config, "parameters", "strategy", "pseudonymize", "hashFunction")
+      .leftMap(_.getMessage)
+
+  private def matchesSchema(config: JValue, schemaKey: SchemaKey): Validation[String, JValue] =
+    if (supportedSchema.matches(schemaKey)) {
+      config.success
+    } else {
+      "Schema key %s is not supported. A '%s' enrichment must have schema '%s'."
+        .format(schemaKey, supportedSchema.name, supportedSchema)
+        .failure
+    }
+}
+
+/**
+ * The PiiPseudonymizerEnrichment runs after all other enrichments to find fields that are configured as PII (personally
+ * identifiable information) and apply some anonymization (currently only psudonymizantion) on them. Currently a single
+ * strategy for all the fields is supported due to the config format, and there is only one implemented strategy,
+ * however the enrichment supports a strategy per field.
+ *
+ * The user may specify two types of fields POJO or JSON. A POJO field is effectively a scalar field in the
+ * EnrichedEvent, whereas a JSON is a "context" formatted field and it can be wither a scalar in the case of
+ * unstruct_event or an array in the case of derived_events and contexts
+ *
+ * @param fieldList a list of configured PiiFields
+ */
+case class PiiPseudonymizerEnrichment(fieldList: List[PiiField]) extends Enrichment {
+  def transformer(event: EnrichedEvent): Unit = fieldList.foreach(_.transform(event))
+}
+
+/**
+ * Specifies a scalar field in POJO and the strategy that should be applied to it.
+ * @param strategy the strategy that should be applied
+ * @param fieldMutator the field mutator where the strategy will be applied
+ */
+final case class PiiScalar(strategy: PiiStrategy, fieldMutator: PiiConstants.Mutator) extends PiiField {
+  override def applyStrategy(fieldValue: String): String = if (fieldValue != null) strategy.scramble(fieldValue) else null
+}
+
+/**
+ * Specifies a strategy to use, a field mutator where the JSON can be found in the EnrichedEvent POJO, a schema criterion to
+ * discriminate which contexts to apply this strategy to, and a json path within the contexts where this strategy will
+ * be applied (the path may correspond to multiple fields).
+ *
+ * @param strategy the strategy that should be applied
+ * @param fieldMutator the field mutator for the json field
+ * @param schemaCriterion the schema for which the strategy will be applied
+ * @param jsonPath the path where the strategy will be applied
+ */
+final case class PiiJson(strategy: PiiStrategy, fieldMutator: PiiConstants.Mutator, schemaCriterion: SchemaCriterion, jsonPath: String)
+    extends PiiField {
+  implicit val json4sFormats = DefaultFormats
+
+  override def applyStrategy(fieldValue: String): String =
+    if (fieldValue != null)
+      compact(render(parse(fieldValue).transformField {
```

Comment:
I'm worry that this function will be applied to all fields in JSON.

```scala
parseJson("""{"foo": 1, "nested": {"foo": 2}}""").transformField {
  case JField("foo", JInt(x)) => JField("bar", JInt(x + 1))
}
JObject(List(("bar", JInt(2)), ("nested", JObject(List(("bar", JInt(3)))))))
```

So, having a context with `schema` and `data` keys seems a bit dangerous for me now. As far as I can understand it won't transform anything if structure is not exactly as configured. Still - maybe this is wasteful to apply this recursive function to all fields in JSON?

Reply:
What you are saying is true, and another reason for me to not like json4s, however there is a guard in modifyObjectIfSchemaMatches, which is it only applies the modification if the schema matches. The only documented way that I could find on modifying a field does exactly that (in http://json4s.org/) and yes it does modify fields in multiple levels using pattern matching.  Using json4s, I am not sure what is a good alternative to avoid looking at all the fields recursively.

Reply:
NM ^^^

Reply:
```scala
def transformObject(json: JValue, f: PartialFunction[JField, JField]): JValue = json match {
  case JObject(fields) =>
    JObject(fields.map { field => if (f.isDefinedAt(field)) f(field) else field })
  case other => other
}
```